serving/turbine_serving/llm/testing/fake_v1_module.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2024 Advanced Micro Devices, Inc
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+"""Implements a service_v1 compliant module in Python for testing.
+
+This uses a PyModuleInterface to define a fake VmModule that exposes 'prefill_bs{n}'
+and 'decode_bs{n}' such that the call sequence and args/results can be manipulated.
+"""
+
+import numpy as np
+import textwrap
+import threading
+
+from iree.runtime import (  # type: ignore
+    BufferUsage,
+    HalBuffer,
+    HalBufferView,
+    HalDevice,
+    HalElementType,
+    HalFence,
+    MemoryType,
+    PyModuleInterface,
+    VmModule,
+    VmRef,
+)
+
+from ..config import ModelParams
+
+
+def create_fake_module(
+    device: HalDevice, module_name: str, model_params: ModelParams
+) -> VmModule:
+    class ServiceV1Module:
+        def __init__(self, iface):
+            ...
+            print("IFACE:", iface, dir(iface))
+
+        def prefill(
+            self,
+            bs: int,
+            token_ids_ref: VmRef,
+            seq_lens_ref: VmRef,
+            attn_block_indices_ref: VmRef,
+            wait_fence_ref: VmRef,
+            signal_fence_ref: VmRef,
+        ):
+            result_array: np.ndarray = np.ndarray([bs, 1], dtype=np.int32)
+
+            def run():
+                print(f"FAKE_V1_MODULE: PREFILL bs={bs} : WAIT")
+                wait_fence = wait_fence_ref.deref(HalFence)  # type: HalFence
+                signal_fence = signal_fence_ref.deref(HalFence)  # type: HalFence
+                try:
+                    wait_fence.wait()
+                    print("  - READY")
+                    _format_device_buffer_view(
+                        lambda s: print("  token_ids =", s), token_ids_ref
+                    )
+                    _format_device_buffer_view(
+                        lambda s: print("  seq_lens =", s), seq_lens_ref
+                    )
+                    _format_device_buffer_view(
+                        lambda s: print("  attn_block_indices =", s),
+                        attn_block_indices_ref,
+                    )
+
+                    # Async populate.
+                    device_array = result_bv.map().asarray(
+                        result_array.shape, result_array.dtype
+                    )
+                    for i in range(bs):
+                        device_array[i, 0] = i + 1
+
+                    signal_fence.signal()
+                except Exception as e:
+                    signal_fence.fail(str(e))
+
+            threading.Thread(target=run).start()
+
+            result_buffer = device.allocator.allocate_buffer(
```

Comment:
if you have it exposed you'll want this to be an async alloca (and if not, would be worth getting that exposed)
any alloc you can make async the better - sync deallocate is still fine, but the best is to also get async dealloca so that we can reclaim the memory ASAP
(we need an ABI attribute to let the compiler know when it can async dealloca any inputs to a function, and by default it always returns async alloca'ed results, and when we line that up right it means in the steady state we shouldn't be doing any sync alloc/dealloc)

Reply:
Thanks - this is just a mock and the real thing pools buffers too. But point taken.